.doom.d/config.org
#+TITLE: Pretty DOOM Configuration
#+PROPERTY: header-args
#+STARTUP: content

* General Settings
** General Stuff
#+begin_src elisp :tangle yes

;; Font - Still not really sure if this overrides nano stuff
(setq doom-font "Iosevka Extended-12"
      doom-variable-pitch-font "Iosevka Extended-13")

;; Theme - Palenight is the classic, sometimes I like a light theme
;; (setq doom-theme 'doom-palenight)
(setq doom-theme 'adwaita)

;; (setq doom-theme 'kaolin-valley-light)

;; Ispell dicitonary
(setq ispell-dictionary "en")
#+end_src

** Other pleasant settings
#+begin_src elisp :tangle yes

;; User name & stuff
(setq user-full-name "Nilay"
      user-mail-address "nilaypatel2@gmail.com")

;; Relative line Numbers (visual makes it account for big lines)
(setq display-line-numbers-type nil)

;; Turn off highlighting on search
(setq evil-ex-search-persistent-highlight nil)
#+end_src

** Disable some stuff
#+begin_src elisp :tangle yes
(setq projectile-track-known-projects-automatically nil)
#+end_src

** NΛNO Bastardization
*** Load the files
#+begin_src elisp :tangle yes
(add-load-path! "/home/offendo/.emacs.d/.local/straight/build-27.1/nano-emacs")
;; this will set all the necessary colors
;; just make sure not to call (nano-theme)
(require 'nano-theme-light)
;: Mandatory
(require 'nano-base-colors)
(require 'nano-faces)
;; Theme -- Not sure if I need this anymore
(require 'nano-theme)
;; Nano header & mode lines
(setq-default mode-line-format nil)
(setq mode-line-format nil)
#+end_src

*** Copy/Paste'd nano-layout
#+begin_src elisp :tangle yes
(setq default-frame-alist
      (append (list
               '(min-height . 1)  '(height     . 90)
               '(min-width  . 1) '(width      . 162)
               '(vertical-scroll-bars . nil)
               '(internal-border-width . 48)
               '(left-fringe    . 0)
               '(right-fringe   . 0)
               '(tool-bar-lines . 0)
               '(menu-bar-lines . 0))))

(defface fallback '((t :family "Fira Code"
                       :inherit 'nano-face-faded)) "Fallback")
(set-display-table-slot standard-display-table 'truncation
                        (make-glyph-code ?… 'fallback))
(set-display-table-slot standard-display-table 'wrap
                        (make-glyph-code ?↩ 'fallback))

(set-fontset-font t nil "Fira Code" nil 'append)

(setq inhibit-startup-screen t
      inhibit-startup-message t
      inhibit-startup-echo-area-message t
      initial-scratch-message nil)
(tool-bar-mode 0)
(tooltip-mode 0)
(menu-bar-mode 0)
(global-hl-line-mode 1)
(setq x-underline-at-descent-line t)

;; Vertical window divider
(setq window-divider-default-right-width 24)
(setq window-divider-default-places 'right-only)
(window-divider-mode nil)

;; No ugly button for checkboxes
(setq widget-image-enable nil)

;; Hide org markup for README
(setq org-hide-emphasis-markers t)
#+end_src

*** Copy/Paste'd nano-writer
#+begin_src elisp :tangle yes
(require 'nano-writer)
#+end_src

*** Copy/Paste'd nano-modeline
#+begin_src elisp :tangle yes
;; not sure why this works and the other one doesn't...
(add-load-path! "/home/offendo/.doom.d/")
(require 'fndo-nano-modeline)
#+end_src

#+end_src
* Evil Mode stuff
** Lazy Load custom packages
#+begin_src elisp :tangle yes
(use-package evil-matchit
  :after evil
  :config
  (global-evil-matchit-mode t))
  #+end_src

** Settings
#+begin_src elisp :tangle yes
;; disable evil-snipe
(after! evil-snipe
  (evil-snipe-mode -1))
#+end_src
** A few classic Vim bindings
#+begin_src elisp :tangle yes
(defmacro define-and-bind-text-object (key start-regex end-regex)
  (let ((inner-name (make-symbol "inner-name"))
        (outer-name (make-symbol "outer-name")))
    `(progn
       (evil-define-text-object ,inner-name (count &optional beg end type)
         (evil-select-paren ,start-regex ,end-regex beg end type count nil))
       (evil-define-text-object ,outer-name (count &optional beg end type)
         (evil-select-paren ,start-regex ,end-regex beg end type count t))
       (define-key evil-inner-text-objects-map ,key (quote ,inner-name))
       (define-key evil-outer-text-objects-map ,key (quote ,outer-name)))))
;; create "il"/"al" (inside/around) line text objects:
(define-and-bind-text-object "l" "^\\s-*" "\\s-*$")
;; create "id"/"ad" (inside/around) entire buffer text objects:
(define-and-bind-text-object "d" "\\`\\s-*" "\\s-*\\'")
                                        ; between dollar signs:
(define-and-bind-text-object "$" "\\$" "\\$")
#+end_src

#+RESULTS:
: outer-name

* Org Mode
** Basics
#+begin_src elisp :tangle yes
;; Somewhere to dump org files like agenda, etc.
(setq org-directory "~/org/")

;; Make the agenda show the whole week starting from today
(after! org-agenda
  (setq org-agenda-span 'week)
  (setq org-agenda-start-day "+0d"))

;; Some nice visual things
(after! org-mode
  (setq org-image-actual-width nil))

;; max column width
(add-hook 'org-mode-hook
          (lambda ()
            (auto-fill-mode)))
#+end_src

** Export
#+begin_src elisp :tangle yes
;; LaTeX Exporting
(after! ox-latex
  ;; latexmk is the only way to go
  (setq org-latex-pdf-process
        (list "latexmk -outdir=./build/ -pdflatex='lualatex -shell-escape -interaction nonstopmode' -pdf -f  %f"))
  ;; Pretty minted stuff
  (setq org-latex-listings 'minted)
  ;; Beamer level should be 2 so we can have sections, e.g. Introduction, Analysis
  (setq org-beamer-frame-level 2)
  ;; make previews normal sized
  (plist-put org-format-latex-options :scale 3)
  ;; don't bother starting up with latex previews, it just makes things slow
  (setq org-startup-with-latex-preview nil)
  ;; latex exporting
  (push '("" "minted" t) org-latex-packages-alist)
  (push '("" "booktabs" t) org-latex-packages-alist)
  (setq org-latex-caption-above nil)
  (push '("homework" "\\documentclass{homework}") org-latex-classes)
  ;; load org-mode bibtex
  (require 'ox-bibtex))
#+end_src

** Jekyll
#+begin_src elisp :tangle yes
;; This lets you publish a "project" via org-mode export:
;; (setq org-publish-project-alist
;;       '(;; This defines how to publish the org-mode content:
;;         ("org-fndo"
;;          ;; Path to your org files.
;;          :base-directory "~/Documents/code/offendo.github.io/org/"
;;          :base-extension "org"
;;
;;          ;; Path to your Jekyll project.
;;          :publishing-directory "~/Documents/code/offendo.github.io/deploy/"
;;          :recursive t
;;          :publishing-function org-html-publish-to-html
;;          :headline-levels 4
;;          :html-extension "html"
;;          :body-only t ;; Only export section between <body> </body>
;;          :with-toc nil
;;          :auto-preamble nil
;;          )
;;         ;; This defines how to publish non-org-mode content e.g. txt files
;;         ("org-static-fndo"
;;          :base-directory "~/Documents/code/offendo.github.io/org/"
;;          :base-extension "css\\|js\\|png\\|jpg\\|gif\\|pdf\\|mp3\\|ogg\\|swf\\|php"
;;          :publishing-directory "~/Documents/code/offendo.github.io/deploy/"
;;          :recursive t
;;          :publishing-function org-publish-attachment)
;;
;;         ;; This says "jb.com" is both of the above:
;;         ("offendo.github.io" :components ("org-fndo" "org-static-fndo"))
;;         ))
;;
;; (defun org-jekyll-post-link-follow (path)
;;   (org-open-file-with-emacs path))
;;
;; (defun org-jekyll-post-link-export (path desc format)
;;   (cond
;;    ((eq format 'html)
;;     (format "<a href=\"{%% post_url %s %%}\">%s</a>" (file-name-sans-extension path) desc))))
;;
;; (org-add-link-type "jekyll-post" 'org-jekyll-post-link-follow 'org-jekyll-post-link-export)
#+end_src

* LSP & Autocompletion
** Company
#+begin_src elisp :tangle yes
;; autocompletion
(use-package! company
  :config (setq company-idle-delay 0.2
                company-minimum-prefix-length 3))

(map! (:map company-mode-map
       :desc "Trigger autocompletion"
       :i "C-l" #'company-search-candidates))
#+end_src

** LSP
#+begin_src elisp :tangle yes

;; LSP mode bindings
(require 'evil)
(map! (:map lsp-mode-map
       :desc "Peek at implementation"
       :nv "gd" #'lsp-ui-peek-find-definitions
       :desc "Peek at references"
       :nv "gr" #'lsp-ui-peek-find-references))

;; Enable the docs
(add-hook 'lsp-mode-hook
          (lambda ()
            (require 'lsp-ui)
            (lsp-ui-mode)
            (setq lsp-ui-doc-max-height 16
                  lsp-ui-doc-max-width 100
                  lsp-ui-doc-enable t ;; disable for now to see how I like it
                  lsp-ui-sideline-ignore-duplicate t
                  lsp-signature-auto-activate nil
                  lsp-enable-symbol-highlighting nil
                  lsp-ui-doc-position 'top)
            ;; (set-face-background 'lsp-ui-doc-background "#292D3E")
            ;; (set-face-background 'child-frame-border "#95a4e1")
            ;; (set-face-background 'lsp-ui-doc-background "#292D3E")
            ;; (set-face-background 'lsp-ui-peek-peek "#292D3E")
            ;; (set-face-background 'lsp-ui-peek-list "#292D3E")
            ))

;; only update checks on save
(setq flycheck-check-syntax-automatically '(mode-enabled save))
#+end_src

* Language Specifics
** Haskell
#+begin_src elisp :tangle yes
;; Haskell Specific stuff
(add-hook 'haskell-mode-hook
          (lambda ()
            (setq lsp-haskell-diagnostics-on-change nil
                  lsp-haskell-hlint-on t)))
#+end_src

** LaTeX
#+begin_src elisp :tangle yes
;; make the column width fixed
(add-hook 'LaTeX-mode-hook
          (lambda ()
            (auto-fill-mode)
            (setq tex-fontify-script nil)
            (setq font-latex-fontify-script nil)
            (setq +latex-viewers '(skim atril evince sumatrapdf zathura okular pdf-tools)
                  TeX-view-program-selection '((output-pdf "Atril")))))
#+end_src

* Email
** mu4e
#+begin_src elisp :tangle yes
(add-load-path! "/usr/local/share/emacs/site-lisp/mu4e/")
(use-package! mu4e
  :load-path "/usr/local/share/emacs/site-lisp/mu4e/"
  :ensure nil
  :config
  (setq mu4e-change-filenames-when-moving t)
  ;; update every 10 minutes
  (setq mu4e-update-interval (* 3 60))
  (setq mu4e-get-mail-command "mbsync -a")
  ;; mail in the mail directory
  (setq mu4e-maildir-list (list "~/mail"))
  ;; set all the different folders
  (setq mu4e-drafts-folder "/[Gmail]/Drafts")
  (setq mu4e-sent-folder   "/[Gmail]/Sent Mail")
  (setq mu4e-refile-folder "/[Gmail]/All Mail")
  (setq mu4e-trash-folder  "/[Gmail]/Trash")
  ;; shortcuts
  (setq mu4e-maildir-shortcuts
        '(("/Inbox" . ?i)
          ("/[Gmail]/Sent Mail" . ?s)
          ("/[Gmail]/All Mail" . ?a)
          ("/[Gmail]/Trash" . ?t)
          ("/[Gmail]/Drafts" . ?d)))
  ;; read html better
  ;; (setq mu4e-html2text-command "html2text")
  ;; send a notification for new emails
  (add-hook 'mu4e-index-updated-hook
            (lambda ()
              (shell-command
               "notify-send -t 5000 -a 'mu4e' -i 'emacs' 'You have new emails\!'"))))
#+end_src


.emacs.d/config.org
#+TITLE: Nilay's Emacs Configuration
#+AUTHOR: Nilay
#+DESCRIPTION: Trying out a non-doom configuration
#+STARTUP: content
#+PROPERTY: header-args :results silent

* Important Packages
** Use-package
#+begin_src elisp
(straight-use-package 'use-package) 
#+end_src

** Evil Mode
Anything related to the fantastic evil-mode. Primarily extensions and stuff.
#+begin_src elisp
(use-package evil
    :init
    (setq evil-want-keybinding nil)
    (setq evil-want-integration t)
    :straight t
    :config
    (evil-mode 1)
    (define-key key-translation-map (kbd "ESC") (kbd "C-g")))
(use-package evil-collection ;; make evil available everywhere
    :after evil
    :straight t
    :init (evil-collection-init))
(use-package undo-fu ;; make undo system work better
    :after evil
    :straight t
    :config
    (evil-set-undo-system 'undo-fu)
    (setq evil-want-fine-undo t))
(use-package evil-goggles ;; show what I'm editing in a flash
    :after evil
    :straight t
    :config
    (evil-goggles-mode)
    (setq evil-goggles-pulse t)
    (setq evil-goggles-duration 0.100))
(use-package evil-surround
  :straight t
  :config
  (global-evil-surround-mode 1))
(use-package evil-org
  :straight t
  :after org
  :hook (org-mode . (lambda () evil-org-mode))
  :config
  (require 'evil-org-agenda)
  (evil-org-agenda-set-keys))
(use-package evil-matchit
  :straight t
  :after evil
  :config (global-evil-matchit-mode 1))
#+end_src

** General
Check out the [[https://github.com/noctuid/general.el][general]] page for more information.
#+begin_src elisp
  (use-package general
    :straight t
    :config
    (general-evil-setup t))
#+end_src

* Emacs Stuff
** Startup Time
This doesn't work with early-init though, so it'll take some work if I ever want to use it
#+begin_src elisp
;; Startup time
(use-package esup
  :straight t)
#+end_src

** Settings
#+begin_src elisp
;; org-mode src blocks
;; (setq org-src-tab-acts-natively t)
(setq org-edit-src-content-indentation 0)
;; Ispell dicitonary
(setq ispell-dictionary "en")
;; Expand tab to spaces
(setq indent-tabs-mode nil)
(setq tab-width 2)
;; disable the scratch message stuff
(setq initial-scratch-message nil)
;; add to path
(add-to-list 'exec-path "~/.cargo/bin/")
;; max column width
(setq-default fill-column 100)
;; cache-directory
(setq nilays-cache-dir (concat user-emacs-directory ".cache/"))
;; enable global hl-line mode
(global-hl-line-mode +1)
;; save history
(setq savehist-file (concat nilays-cache-dir "savehist"))
(savehist-mode 1)
;; backup files all in one place
(setq backupdir (concat nilays-cache-dir "backup/"))
(setq backup-directory-alist '(("" . 'backupdir)))
#+end_src

** Basic Keybindings
#+begin_src elisp
(general-create-definer leader-def
  :states '(normal insert emacs)
  :prefix "SPC"
  :non-normal-prefix "M-SPC"
  :prefix-command 'leader-prefix-command
  :prefix-map 'leader-prefix-map)

;; Find-file
(leader-def "." 'find-file)

;; Buffer editor
(leader-def "b" 'ibuffer)

;; Dired
(defun nilay-open-dired ()
  "Open dired"
  (interactive)
  (dired "."))

(leader-def "-" 'nilay-open-dired)

;; Bookmarks
(leader-def "v" 'bookmark-map)
#+end_src

** Text Objects
#+begin_src elisp
(defmacro define-and-bind-text-object (key start-regex end-regex)
  (let ((inner-name (make-symbol "inner-name"))
        (outer-name (make-symbol "outer-name")))
    `(progn
       (evil-define-text-object ,inner-name (count &optional beg end type)
         (evil-select-paren ,start-regex ,end-regex beg end type count nil))
       (evil-define-text-object ,outer-name (count &optional beg end type)
         (evil-select-paren ,start-regex ,end-regex beg end type count t))
       (define-key evil-inner-text-objects-map ,key (quote ,inner-name))
       (define-key evil-outer-text-objects-map ,key (quote ,outer-name)))))


;; i$/a$ objects
(define-and-bind-text-object "$" "\\$" "\\$")

;; create "id"/"ad" (inside/around) entire buffer text objects:
(define-and-bind-text-object "d" "\\`\\s-*" "\\s-*\\'")

(defun nilay-end-of-non-whitespace()
  "Move to the last non-whitespace character in the current line."
  (interactive)
  (move-end-of-line nil)
  (re-search-backward "^\\|[^[:space:]]"))                                       

(evil-define-text-object nilay-inner-line (count &optional beg end type)
  "Select a line ignoring starting and ending whitespace"
  (save-excursion
    (evil-first-non-blank)
    (let ((begp (point)))
      (nilay-end-of-non-whitespace)
      (let  ((endp (point)))
        (evil-range begp (+ 1 endp))))))

(general-define-key
 :keymaps 'evil-inner-text-objects-map
 "l" 'nilay-inner-line)
#+end_src

** Emacs Anywhere
#+begin_src elisp
(setq ea-window-title "floating")
#+end_src

* Language Settings
** LaTeX
#+begin_src elisp
(use-package latex
  :straight nil
  :hook (LaTeX-mode . auto-fill-mode)
  :config
  (setq tex-fontify-script nil
        font-latex-fontify-script nil
        font-latex-fontify-sectioning 'color)
  (setq +latex-viewers '(skim atril evince sumatrapdf zathura okular pdf-tools)
    TeX-view-program-selection '((output-pdf "Atril"))))
(use-package auctex
  :straight t
  :defer t
  :config
  (add-to-list
   'TeX-command-list
   '("LatexMk" "latexmk -outdir=./build/ -pdflatex='lualatex -shell-escape -interaction nonstopmode' -pdf -f  %f"
     TeX-run-command t t
     :help "Run LatexMk"))
  (setq 'TeX-command-default "LatexMk"))
#+end_src

* Cool Things
** Almighty Org
#+begin_src elisp
(use-package org
  :mode (("\\.org$" . org-mode))
  :straight nil
  :hook (org-mode .  auto-fill-mode)
        (org-mode . (lambda () (electric-indent-local-mode -1)))
  :config
  (set-display-table-slot standard-display-table
                          'selective-display (string-to-vector " ↷"))
  (setq org-startup-indented t
        org-confirm-babel-evaluate nil
        org-agenda-span 'week
        org-agenda-start-day "+0d"
        org-image-actual-width nil))
;; Prettify
  (use-package org-superstar
    :straight t
    :hook (org-mode . org-superstar-mode)
    :config
    (org-superstar-configure-like-org-bullets))
#+end_src

** Selectrum et al.
#+begin_src elisp
;; recentf-mode
(use-package recentf
  :after consult
  :config 
  (recentf-mode 1)
  (setq recentf-save-file (concat nilays-cache-dir "recentf" )
        recentf-max-menu-items 10
        recentf-max-saved-items 10))
;; Selectrum
(use-package selectrum
  :straight t
  :config
  (selectrum-mode +1)
  (setq completion-ignore-case t
	read-file-name-completion-ignore-case t
	read-buffer-completion-ignore-case t))
;; Prescient
(use-package prescient
  :straight t)
(use-package selectrum-prescient
  :straight t
  :after '(selectrum prescient)
  :config
  (selectrum-prescient-mode)
  (setq selectrum-prescient-enable-sorting t)
  (setq selectrum-prescient-enable-filtering t)
  (prescient-persist-mode))
;; Orderless
(use-package orderless
  :straight t
  :custom (completion-styles '(orderless substring )))
;; Consult
(use-package consult
  :straight t
  :general
  (leader-def "," 'consult-buffer)
  (leader-def "y" 'consult-register)
  (leader-def "m" 'consult-mark)
  (leader-def "f" 'consult-ripgrep))
;; Marginalia
(use-package marginalia
  :general
  (:keymaps 'minibuffer-local-map
	    "M-A"  'marginalia-cycle)
  :init
  (marginalia-mode)
  (marginalia-cycle)
  (advice-add #'marginalia-cycle
	      :after (lambda ()
		       (when (bound-and-true-p selectrum-mode)
			 (selectrum-exhibit 'keep-selected)))))
;; Embark 
(use-package embark
  :straight t
  :general
  ("C-a" 'embark-act)
  (:keymaps 'embark-file-map
  "d" 'delete-file
  "r" 'rename-file
  "c" 'copy-file
  "v" 'vonfind-file-other-window))

(use-package embark-consult
  :straight t
  :after '(embark consult)
  :demand t
  :hook
  (embark-collect-mode . embark-consult-preview-minor-mode))
#+end_src

** Company
#+begin_src elisp
(use-package company
  :straight t
  :config
  (setq company-idle-delay 0.2 
        company-show-numbers t 
        company-tooltip-limit 10 
        company-minimum-prefix-length 2 
        company-tooltip-align-annotations t
        ;; invert the navigation direction if the the completion popup-isearch-match
        ;; is displayed on top (happens near the bottom of windows)
        company-tooltip-flip-when-above t)
  (global-company-mode 1))
#+end_src

** Magit
#+begin_src elisp
  (use-package magit
    :straight t
    :general
    (leader-def "g" 'magit-status))
#+end_src

** Which-key
#+begin_src elisp
  (use-package which-key
    :straight t
    :config
    (which-key-mode)
    (setq which-key-idle-delay 0.5))
#+end_src

** Yasnippet
#+begin_src elisp
  (use-package yasnippet
    :straight t
    :hook (prog-mode . yas-minor-mode)
          (org-mode . yas-minor-mode)
    :config
    (setq yas-verbosity 1 ; No need to be so verbose
          yas-wrap-around-region t))
  (use-package yasnippet-snippets
    :straight t
    :after yasnippet)
#+end_src

** Formatter
#+begin_src elisp
  (use-package format-all
    :straight t
    :hook (prog-mode . format-all-mode))
#+end_src

** Vterm
#+begin_src elisp
  (use-package vterm
    :straight t)
  (use-package vterm-toggle
    :straight t
    :general (leader-def "t" 'vterm-toggle))
#+end_src

** LSP
#+begin_src elisp
(use-package lsp-mode
  :hook prog-mode
  :straight t)
#+end_src

** Code Folding
#+begin_src elisp
(use-package vimish-fold
  :straight t)
(use-package evil-vimish-fold
  :straight t
  :after '(vimish-fold evil)
  :hook ((prog-mode conf-mode text-mode) . evil-vimish-fold-mode))
(use-package hideshow
  :straight nil
  :hook ((prog-mode org-mode text-mode) . hs-minor-mode))
#+end_src

** Projectile
#+begin_src elisp
(use-package projectile
  :straight t
  :init 
  (setq projectile-cache-file (concat nilays-cache-dir ".projectile.cache"))
  (projectile-mode +1)
  (setq projectile-auto-discover nil)
  :general
  (leader-def "p" 'projectile-command-map)
  (:keymaps 'projectile-command-map
  "d" 'projectile-remove-known-project
  "a" 'projectile-add-known-project
))
#+end_src

* Prettifying
** Adjust sizes
#+begin_src elisp
(setq default-frame-alist
      (append (list
              '(vertical-scroll-bars . nil)
              '(internal-border-width . 24)
              '(left-fringe    . 1)
              '(right-fringe   . 1)
              '(tool-bar-lines . 0)
              '(menu-bar-lines . 0))))

(defun nilays-set-fonts ()
    (set-face-attribute 'variable-pitch nil :font "Iosevka Medium Extended-26")
    (set-face-attribute 'fixed-pitch nil :font "Iosevka Medium Extended-22")
    (set-face-attribute 'default nil :font "Iosevka Medium Extended-22"))
(nilays-set-fonts)
#+end_src

** Modeline
#+begin_src elisp
(use-package doom-modeline
  :straight t
  :init (doom-modeline-mode 1))
#+end_src

** Theme
#+begin_src elisp
;; load a theme
(use-package spacemacs-common
  :straight spacemacs-theme)
;; :init (load-theme 'spacemacs-light t)
#+end_src

** Frame Hook
#+begin_src elisp
(defun nilays-appearance-setup (frame)
  (with-selected-frame frame
    (remove-hook 'after-make-frame-functions 'nilay-appearance-setup)

    (when (functionp 'scroll-bar-mode) (scroll-bar-mode -1))
    (when (functionp 'tool-bar-mode) (tool-bar-mode -1))
    (when (functionp 'menu-bar-mode) (menu-bar-mode -1))

    (load-theme 'spacemacs-light t)
    (nilays-set-fonts)

    (global-hl-line-mode)

    (setq-default system-time-locale "C")))

(if (daemonp)
    (add-hook 'after-make-frame-functions 'nilays-appearance-setup)
  (nilays-appearance-setup (car (frame-list))))
#+end_src




.doom.d/config.el
;; Font - Still not really sure if this overrides nano stuff
(setq doom-font "Iosevka Extended-12"
      doom-variable-pitch-font "Iosevka Extended-13")

;; Theme - Palenight is the classic, sometimes I like a light theme
;; (setq doom-theme 'doom-palenight)
(setq doom-theme 'adwaita)

;; (setq doom-theme 'kaolin-valley-light)

;; Ispell dicitonary
(setq ispell-dictionary "en")

;; User name & stuff
(setq user-full-name "Nilay"
      user-mail-address "nilaypatel2@gmail.com")

;; Relative line Numbers (visual makes it account for big lines)
(setq display-line-numbers-type nil)

;; Turn off highlighting on search
(setq evil-ex-search-persistent-highlight nil)

(setq projectile-track-known-projects-automatically nil)

(add-load-path! "/home/offendo/.emacs.d/.local/straight/build-27.1/nano-emacs")
;; this will set all the necessary colors
;; just make sure not to call (nano-theme)
(require 'nano-theme-light)
;: Mandatory
(require 'nano-base-colors)
(require 'nano-faces)
;; Theme -- Not sure if I need this anymore
(require 'nano-theme)
;; Nano header & mode lines
(setq-default mode-line-format nil)
(setq mode-line-format nil)

(setq default-frame-alist
      (append (list
               '(min-height . 1)  '(height     . 90)
               '(min-width  . 1) '(width      . 162)
               '(vertical-scroll-bars . nil)
               '(internal-border-width . 48)
               '(left-fringe    . 0)
               '(right-fringe   . 0)
               '(tool-bar-lines . 0)
               '(menu-bar-lines . 0))))

(defface fallback '((t :family "Fira Code"
                       :inherit 'nano-face-faded)) "Fallback")
(set-display-table-slot standard-display-table 'truncation
                        (make-glyph-code ?… 'fallback))
(set-display-table-slot standard-display-table 'wrap
                        (make-glyph-code ?↩ 'fallback))

(set-fontset-font t nil "Fira Code" nil 'append)

(setq inhibit-startup-screen t
      inhibit-startup-message t
      inhibit-startup-echo-area-message t
      initial-scratch-message nil)
(tool-bar-mode 0)
(tooltip-mode 0)
(menu-bar-mode 0)
(global-hl-line-mode 1)
(setq x-underline-at-descent-line t)

;; Vertical window divider
(setq window-divider-default-right-width 24)
(setq window-divider-default-places 'right-only)
(window-divider-mode nil)

;; No ugly button for checkboxes
(setq widget-image-enable nil)

;; Hide org markup for README
(setq org-hide-emphasis-markers t)

(require 'nano-writer)

;; not sure why this works and the other one doesn't...
(add-load-path! "/home/offendo/.doom.d/")
(require 'fndo-nano-modeline)

(use-package evil-matchit
  :after evil
  :config
  (global-evil-matchit-mode t))

;; disable evil-snipe
(after! evil-snipe
  (evil-snipe-mode -1))

(defmacro define-and-bind-text-object (key start-regex end-regex)
  (let ((inner-name (make-symbol "inner-name"))
        (outer-name (make-symbol "outer-name")))
    `(progn
       (evil-define-text-object ,inner-name (count &optional beg end type)
         (evil-select-paren ,start-regex ,end-regex beg end type count nil))
       (evil-define-text-object ,outer-name (count &optional beg end type)
         (evil-select-paren ,start-regex ,end-regex beg end type count t))
       (define-key evil-inner-text-objects-map ,key (quote ,inner-name))
       (define-key evil-outer-text-objects-map ,key (quote ,outer-name)))))
;; create "il"/"al" (inside/around) line text objects:
(define-and-bind-text-object "l" "^\\s-*" "\\s-*$")
;; create "id"/"ad" (inside/around) entire buffer text objects:
(define-and-bind-text-object "d" "\\`\\s-*" "\\s-*\\'")
                                        ; between dollar signs:
(define-and-bind-text-object "$" "\\$" "\\$")

;; Somewhere to dump org files like agenda, etc.
(setq org-directory "~/org/")

;; Make the agenda show the whole week starting from today
(after! org-agenda
  (setq org-agenda-span 'week)
  (setq org-agenda-start-day "+0d"))

;; Some nice visual things
(after! org-mode
  (setq org-image-actual-width nil))

;; max column width
(add-hook 'org-mode-hook
          (lambda ()
            (auto-fill-mode)))

;; LaTeX Exporting
(after! ox-latex
  ;; latexmk is the only way to go
  (setq org-latex-pdf-process
        (list "latexmk -outdir=./build/ -pdflatex='lualatex -shell-escape -interaction nonstopmode' -pdf -f  %f"))
  ;; Pretty minted stuff
  (setq org-latex-listings 'minted)
  ;; Beamer level should be 2 so we can have sections, e.g. Introduction, Analysis
  (setq org-beamer-frame-level 2)
  ;; make previews normal sized
  (plist-put org-format-latex-options :scale 3)
  ;; don't bother starting up with latex previews, it just makes things slow
  (setq org-startup-with-latex-preview nil)
  ;; latex exporting
  (push '("" "minted" t) org-latex-packages-alist)
  (push '("" "booktabs" t) org-latex-packages-alist)
  (setq org-latex-caption-above nil)
  (push '("homework" "\\documentclass{homework}") org-latex-classes)
  ;; load org-mode bibtex
  (require 'ox-bibtex))

;; This lets you publish a "project" via org-mode export:
;; (setq org-publish-project-alist
;;       '(;; This defines how to publish the org-mode content:
;;         ("org-fndo"
;;          ;; Path to your org files.
;;          :base-directory "~/Documents/code/offendo.github.io/org/"
;;          :base-extension "org"
;;
;;          ;; Path to your Jekyll project.
;;          :publishing-directory "~/Documents/code/offendo.github.io/deploy/"
;;          :recursive t
;;          :publishing-function org-html-publish-to-html
;;          :headline-levels 4
;;          :html-extension "html"
;;          :body-only t ;; Only export section between <body> </body>
;;          :with-toc nil
;;          :auto-preamble nil
;;          )
;;         ;; This defines how to publish non-org-mode content e.g. txt files
;;         ("org-static-fndo"
;;          :base-directory "~/Documents/code/offendo.github.io/org/"
;;          :base-extension "css\\|js\\|png\\|jpg\\|gif\\|pdf\\|mp3\\|ogg\\|swf\\|php"
;;          :publishing-directory "~/Documents/code/offendo.github.io/deploy/"
;;          :recursive t
;;          :publishing-function org-publish-attachment)
;;
;;         ;; This says "jb.com" is both of the above:
;;         ("offendo.github.io" :components ("org-fndo" "org-static-fndo"))
;;         ))
;;
;; (defun org-jekyll-post-link-follow (path)
;;   (org-open-file-with-emacs path))
;;
;; (defun org-jekyll-post-link-export (path desc format)
;;   (cond
;;    ((eq format 'html)
;;     (format "<a href=\"{%% post_url %s %%}\">%s</a>" (file-name-sans-extension path) desc))))
;;
;; (org-add-link-type "jekyll-post" 'org-jekyll-post-link-follow 'org-jekyll-post-link-export)

;; autocompletion
(use-package! company
  :config (setq company-idle-delay 0.2
                company-minimum-prefix-length 3))

(map! (:map company-mode-map
       :desc "Trigger autocompletion"
       :i "C-l" #'company-search-candidates))

;; LSP mode bindings
(require 'evil)
(map! (:map lsp-mode-map
       :desc "Peek at implementation"
       :nv "gd" #'lsp-ui-peek-find-definitions
       :desc "Peek at references"
       :nv "gr" #'lsp-ui-peek-find-references))

;; Enable the docs
(add-hook 'lsp-mode-hook
          (lambda ()
            (require 'lsp-ui)
            (lsp-ui-mode)
            (setq lsp-ui-doc-max-height 16
                  lsp-ui-doc-max-width 100
                  lsp-ui-doc-enable t ;; disable for now to see how I like it
                  lsp-ui-sideline-ignore-duplicate t
                  lsp-signature-auto-activate nil
                  lsp-enable-symbol-highlighting nil
                  lsp-ui-doc-position 'top)
            ;; (set-face-background 'lsp-ui-doc-background "#292D3E")
            ;; (set-face-background 'child-frame-border "#95a4e1")
            ;; (set-face-background 'lsp-ui-doc-background "#292D3E")
            ;; (set-face-background 'lsp-ui-peek-peek "#292D3E")
            ;; (set-face-background 'lsp-ui-peek-list "#292D3E")
            ))

;; only update checks on save
(setq flycheck-check-syntax-automatically '(mode-enabled save))

;; Haskell Specific stuff
(add-hook 'haskell-mode-hook
          (lambda ()
            (setq lsp-haskell-diagnostics-on-change nil
                  lsp-haskell-hlint-on t)))

;; make the column width fixed
(add-hook 'LaTeX-mode-hook
          (lambda ()
            (auto-fill-mode)
            (setq tex-fontify-script nil)
            (setq font-latex-fontify-script nil)
            (setq +latex-viewers '(skim atril evince sumatrapdf zathura okular pdf-tools)
                  TeX-view-program-selection '((output-pdf "Atril")))))

(add-load-path! "/usr/local/share/emacs/site-lisp/mu4e/")
(use-package! mu4e
  :load-path "/usr/local/share/emacs/site-lisp/mu4e/"
  :ensure nil
  :config
  (setq mu4e-change-filenames-when-moving t)
  ;; update every 10 minutes
  (setq mu4e-update-interval (* 3 60))
  (setq mu4e-get-mail-command "mbsync -a")
  ;; mail in the mail directory
  (setq mu4e-maildir-list (list "~/mail"))
  ;; set all the different folders
  (setq mu4e-drafts-folder "/[Gmail]/Drafts")
  (setq mu4e-sent-folder   "/[Gmail]/Sent Mail")
  (setq mu4e-refile-folder "/[Gmail]/All Mail")
  (setq mu4e-trash-folder  "/[Gmail]/Trash")
  ;; shortcuts
  (setq mu4e-maildir-shortcuts
        '(("/Inbox" . ?i)
          ("/[Gmail]/Sent Mail" . ?s)
          ("/[Gmail]/All Mail" . ?a)
          ("/[Gmail]/Trash" . ?t)
          ("/[Gmail]/Drafts" . ?d)))
  ;; read html better
  ;; (setq mu4e-html2text-command "html2text")
  ;; send a notification for new emails
  (add-hook 'mu4e-index-updated-hook
            (lambda ()
              (shell-command
               "notify-send -t 5000 -a 'mu4e' -i 'emacs' 'You have new emails\!'"))))
